cluster: support automatically assigning a node ID

andrwng opened the pull request:
## Cover letter

This PR contains functionality for automatically assigning a node ID. For details of the startup changes, see the comments in `cluster_discovery.h` and `members_manager.h`.

This PR adds the ability to start a node without explicitly assigning its node ID. When a node starts up, before it can initialize most of its subsystems, it must be made aware of its node ID. It can either get this from its config, the kv-store, or be assigned one by the controller leader. In all cases, the node ID and node UUID must be registered with the controller, after which the node can proceed join the cluster.

The high level steps are as follows:

When the node ID is unknown:
1. Generate or load node UUID
2. Get assigned a node ID by sending a request to the controller leader
3. Start subsystems with our known node ID
4. Join the cluster and get added to the controller Raft group by sending a request to the controller leader
5. Once added to the cluster, open endpoints for user traffic

When the node ID is known:
1. Generate or load node UUID
2. Load node ID from config or kv-store
3. Start subsystems with our known node ID
4. Register our UUID with our node ID and join the cluster by sending a request to the controller leader
5. Once added to the cluster, open endpoints for user traffic

The registration of the UUID is handled by the existing join_node RPC, which has a currently unused field for node UUID. Based on whether the registration has already occurred, the `members_manager` will either register the UUID (potentially assigning a node ID in doing so), persisting it to the controller log, or simply proceed to add the node to the cluster. Node DI assignments are made deterministic by having node ID assignment and registration be done in the apply phase of a new controller command.


## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [x] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes

Users are no longer required to supply a node ID in each node's node configuration file.

<!-- don't ship user breaking changes. Ping PMs for help with user visible changes  -->

## Release notes

* Users are no longer required to supply a node ID in each node's node configuration file. With such a configuration, Redpanda assigns a node ID to each empty node as it joins the cluster.

### Features

* Users are no longer required to supply a node ID in each node's node configuration file. All nodes must be upgraded before using this feature. Node IDs on existing nodes will be preserved when using this.

andrwng commented:
/ci-repeat 10

andrwng commented on src/v/redpanda/application.cc:
```diff
@@ -1391,6 +1392,12 @@ void application::wire_up_and_start(::stop_signal& app_signal, bool test_mode) {
+    if (config::node().node_id() == std::nullopt) {
+        // If we previously didn't have a node ID, set it in the config. We
+        // will persist it in the kvstore when the controller starts up.
+        config::node().node_id.set_value(
```
Need to call this on all cores.

mmaslankaprv commented on src/v/cluster/members_manager.cc:
```diff
@@ -423,6 +451,49 @@ void members_manager::join_raft0() {
+bool members_manager::try_register_node_id(
+  const model::node_id& requested_node_id,
+  const model::node_uuid& requested_node_uuid) {
+    vassert(requested_node_id != model::node_id(-1), "invalid node ID");
```
should we check `requested_node_id >= 0` ?

mmaslankaprv commented on src/v/cluster/members_manager.cc:
```diff
@@ -423,6 +451,49 @@ void members_manager::join_raft0() {
+    vassert(requested_node_id != model::node_id(-1), "invalid node ID");
+    const auto it = _id_by_uuid.find(requested_node_uuid);
+    if (it == _id_by_uuid.end()) {
+        if (_members_table.local().contains(requested_node_id)) {
```
just to make sure if i understand this correctly, if a node running new redpanda version that was already part of the cluster and have node_id set in configuration file will have a data directory wiped it will regenerate UUID and will be remapped to new UUID ?

andrwng replied:
If you don't mind, I will do this in a follow-up change, so we can merge this sooner.

mmaslankaprv commented on src/v/cluster/members_manager.cc:
```diff
@@ -270,6 +274,30 @@ members_manager::apply_update(model::record_batch b) {
+            clusterlog.info,
+            "Applying register_node_uuid_cmd for UUID {}",
+            node_uuid);
+          auto node_id = get_or_assign_node_id(node_uuid);
```
should we validate result here ? node_id may in rare cases be an empty optional

mmaslankaprv commented on src/v/cluster/members_manager.h:
```diff
@@ -190,7 +230,11 @@ class members_manager {
     const std::chrono::milliseconds _join_timeout;
     const consensus_ptr _raft0;
+
+    ss::sharded<controller_stm>& _controller_stm;
```
woulld it be possible not to depend on the `_controller_stm` here ? It creates circular dependency that will be hard to untangle.

andrwng replied:
Current thought is to maybe have this depend on `members_frontend`, which doesn't feel great because the call we need (replicate on the controller log right now) is a bit different from the current frontend pattern (dispatch to leader, potentially replicating to the controller log).

I left this as is for now, but may revisit if you feel strongly about it / it causes trouble.

mmaslankaprv commented on src/v/redpanda/application.cc:
```diff
@@ -23,6 +23,7 @@
 #include "cluster/fwd.h"
 #include "cluster/id_allocator.h"
 #include "cluster/id_allocator_frontend.h"
+#include "cluster/members_manager.h"
```
nit: not needed include

mmaslankaprv commented on src/v/cluster/members_manager.cc:
```diff
@@ -278,6 +283,10 @@ members_manager::apply_update(model::record_batch b) {
       [this](register_node_uuid_cmd cmd) {
           const auto& node_uuid = cmd.key;
           const auto& requested_node_id = cmd.value;
+          vlog(
+            clusterlog.info,
+            "Applying register_node_uuid_cmd for UUID {}",
+            node_uuid);
```
nit: this should be a part of a commit introducing handling for `register_node_uuid_cmd`

mmaslankaprv commented on src/v/cluster/members_manager.cc:
```diff
@@ -455,6 +473,11 @@ bool members_manager::try_register_node_id(
   const model::node_id& requested_node_id,
   const model::node_uuid& requested_node_uuid) {
     vassert(requested_node_id != model::node_id(-1), "invalid node ID");
+    vlog(
+      clusterlog.info,
+      "Registering node ID {} as node UUID {}",
+      requested_node_id,
+      requested_node_uuid);
```
part of previous commit ?

andrwng commented:
Force pushes:
- Addressing feedback: https://github.com/redpanda-data/redpanda/compare/cd3e2f07a466f5f0c155a60f407545861e7c009c..1b06853979191714ee5caa047b52647bf28691e7
- Rebase: https://github.com/redpanda-data/redpanda/compare/1b06853979191714ee5caa047b52647bf28691e7..735af042f011a3fae11b9765fd291f530314d517
- Lint: https://github.com/redpanda-data/redpanda/compare/735af042f011a3fae11b9765fd291f530314d517..abd9cf4ad9e6ad76dad1ecbb8750d00321cf7244

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -43,6 +68,60 @@ vector<broker> cluster_discovery::initial_raft0_brokers() const {
+        co_return ss::stop_iteration::yes;
+    }
+    co_await ss::sleep_abortable(_join_retry_jitter.next_duration(), _as);
+    co_return ss::stop_iteration::no;
```
Return booleans.

andrwng replied:
Done

andrwng commented on src/v/cluster/cluster_discovery.cc:
```diff
@@ -43,6 +68,60 @@ vector<broker> cluster_discovery::initial_raft0_brokers() const {
+              std::current_exception());
+            continue;
+        }
+        if (!r || r.has_error() || !r.value().success) {
+            vlog(
+              clusterlog.debug,
+              "Error registering UUID {}, retrying",
+              _node_uuid);
+            continue;
```
nit: maybe also print the error if there is one

andrwng replied:
Done

andrwng commented on src/v/cluster/members_manager.cc:
```diff
@@ -570,92 +584,211 @@ auto members_manager::dispatch_rpc_to_leader(
+      node_uuid,
+      errc);
+    if (errc != errc::success) {
+        co_return ret_t(join_node_reply{false, model::node_id(-1)});
```
Lost in the rebase -- return errc.

andrwng replied:
Done

andrwng commented on src/v/cluster/members_manager.cc:
```diff
@@ -570,92 +584,211 @@ auto members_manager::dispatch_rpc_to_leader(
+          node_uuid,
+          *node_id,
+          assigned_node_id);
+        co_return ret_t(join_node_reply{false, model::node_id(-1)});
```
Return errc.

andrwng replied:
Done

andrwng commented on src/v/cluster/members_manager.cc:
```diff
@@ -756,15 +756,6 @@ members_manager::handle_join_request(join_node_request const req) {
-          req.node.id(),
-          req.node.rpc_address());
-        co_return ret_t(join_node_reply{false, model::node_id(-1)});
-    }
```
Guard this with `node_id_assignment_supported`

andrwng replied:
Done

emaxerrno commented:
@andrwng this is such a large and impactful change w.r.t public API that would love to attach a .md with the result of the internal RFC as a way to explain it to external users and for documentation folk to be able to write a good guide on this as well as migration from previous behavior. Not a blocker to this PR, but we should have it in the repo. If you already have a google doc on this, would be good to convert it to a .md file and attach it to this pr in /docs folder. I expect a lot of content to get updated based on this.

very excited to see simplification of bootstrapping 
 
cc: @Feediver1

Feediver1 commented:
@andrwng When you get a chance, please send us a [doc Issue ](https://github.com/redpanda-data/documentation/issues/new/choose) for this and we will make sure it is documented. If you link to this PR and any other PRD info/guidance available, that would be great!
cc: @micheleRP

andrwng commented:
CI failure was:
- `OSError: [Errno 28] No space left on device`

andrwng commented:
Force push:
- Add handling for overflow
- https://github.com/redpanda-data/redpanda/compare/26a992f624107dfbe00c2a58ca026a9ed1b34c3f..aba167dd14b48070be51136bc97995bd45d4e63e
